Improve performance by moving branch target checks into translation

arichardson opened the pull request:
Overall this reduces the kernel startup from:
```
       1197.128333      task-clock (msec)         #    0.577 CPUs utilized            ( +-  0.92% )
               431      context-switches          #    0.360 K/sec                    ( +-  0.98% )
                 2      cpu-migrations            #    0.002 K/sec                    ( +- 44.44% )
             7,266      page-faults               #    0.006 M/sec                    ( +-  0.77% )
     4,817,409,965      cycles                    #    4.024 GHz                      ( +-  0.30% )  (50.01%)
    13,510,506,710      instructions              #    2.80  insn per cycle           ( +-  0.45% )  (62.61%)
     2,365,632,069      branches                  # 1976.089 M/sec                    ( +-  0.38% )  (62.59%)
         7,593,704      branch-misses             #    0.32% of all branches          ( +-  1.42% )  (62.60%)
     3,485,048,757      L1-dcache-loads           # 2911.174 M/sec                    ( +-  0.35% )  (62.67%)
        29,513,912      L1-dcache-load-misses     #    0.85% of all L1-dcache hits    ( +-  1.01% )  (62.42%)
         2,903,959      LLC-loads                 #    2.426 M/sec                    ( +-  2.46% )  (49.84%)
           319,993      LLC-load-misses           #   11.02% of all LL-cache hits     ( +-  3.77% )  (49.86%)

       2.075448587 seconds time elapsed                                          ( +-  0.79% )
```
to
```
       1124.005361      task-clock (msec)         #    0.554 CPUs utilized            ( +-  0.56% )
               412      context-switches          #    0.367 K/sec                    ( +-  0.99% )
                 5      cpu-migrations            #    0.004 K/sec                    ( +- 23.75% )
             7,220      page-faults               #    0.006 M/sec                    ( +-  0.61% )
     4,479,150,899      cycles                    #    3.985 GHz                      ( +-  0.17% )  (49.70%)
    12,678,267,275      instructions              #    2.83  insn per cycle           ( +-  0.27% )  (62.33%)
     2,207,617,055      branches                  # 1964.063 M/sec                    ( +-  0.18% )  (62.48%)
         7,085,827      branch-misses             #    0.32% of all branches          ( +-  1.05% )  (62.59%)
     3,211,821,304      L1-dcache-loads           # 2857.479 M/sec                    ( +-  0.18% )  (62.83%)
        27,558,280      L1-dcache-load-misses     #    0.86% of all L1-dcache hits    ( +-  0.85% )  (62.73%)
         2,602,280      LLC-loads                 #    2.315 M/sec                    ( +-  2.30% )  (49.91%)
           334,191      LLC-load-misses           #   12.84% of all LL-cache hits     ( +-  2.60% )  (49.76%)

       2.029553619 seconds time elapsed                                          ( +-  0.45% )
```

After this change the remaining hot functions are as follows:
```
Samples: 4K of event 'cycles:ppp', Event count (approx.): 4513576048
  Children      Self  Command          Shared Object                      Symbol
+   13.50%    13.47%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] cheri_tag_invalidate
+   12.27%     0.00%  qemu-system-che  [unknown]                          [.] 0x0000000000000001
+    9.72%     0.03%  qemu-system-che  [unknown]                          [.] 0000000000000000
+    8.48%     8.45%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] check_ddc
+    7.26%     7.26%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] address_space_translate_internal
+    6.30%     6.24%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] liveness_pass_1
+    6.29%     6.29%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] get_physical_address
+    5.94%     0.61%  qemu-system-che  libc-2.27.so                       [.] __clock_gettime
+    5.49%     1.77%  qemu-system-che  [vdso]                             [.] __vdso_clock_gettime
+    3.89%     3.89%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] cpu_exec
+    3.70%     3.70%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] r4k_map_address
+    3.21%     3.19%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] tcg_optimize
     2.92%     2.92%  qemu-system-che  [vdso]                             [.] 0x0000000000000977
+    2.17%     2.10%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] tb_gen_code
+    2.16%     2.12%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] helper_lookup_tb_ptr
+    1.86%     1.86%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] tb_find
+    1.71%     1.24%  qemu-system-che  libc-2.27.so                       [.] __memmove_avx_unaligned_erms
+    1.64%     1.62%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] tcg_reg_alloc_op
+    1.55%     1.55%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] helper_ddc_check_load
+    1.21%     1.21%  qemu-system-che  qemu-system-cheri128.btarget_omit  [.] qemu_clock_get_ns
```

The easiest win now is probably caching tag address lookups and saving DDC base+top+tag in DisasContext and retranslating if it changes.

arichardson committed:
Move a PCC bounds assertion to common CHERI code

arichardson committed:
Minor refactoring to share more code

Some ugly cyclic dependencies workarounds to reuse helpers inside accel/tcg/translator.c

arichardson committed:
[CHERI-MIPS] Perform PCC bounds checks for branches in translator

IPS go from ~84M to ~90M and cycles from 4.8bn to 4.5bn when starting
the CheriBSD kernel. Register branch checks (jr) are about 0.4% of the
total runtime so I'll address that next.

Before:
       1197.128333      task-clock (msec)         #    0.577 CPUs utilized            ( +-  0.92% )
               431      context-switches          #    0.360 K/sec                    ( +-  0.98% )
                 2      cpu-migrations            #    0.002 K/sec                    ( +- 44.44% )
             7,266      page-faults               #    0.006 M/sec                    ( +-  0.77% )
     4,817,409,965      cycles                    #    4.024 GHz                      ( +-  0.30% )  (50.01%)
    13,510,506,710      instructions              #    2.80  insn per cycle           ( +-  0.45% )  (62.61%)
     2,365,632,069      branches                  # 1976.089 M/sec                    ( +-  0.38% )  (62.59%)
         7,593,704      branch-misses             #    0.32% of all branches          ( +-  1.42% )  (62.60%)
     3,485,048,757      L1-dcache-loads           # 2911.174 M/sec                    ( +-  0.35% )  (62.67%)
        29,513,912      L1-dcache-load-misses     #    0.85% of all L1-dcache hits    ( +-  1.01% )  (62.42%)
         2,903,959      LLC-loads                 #    2.426 M/sec                    ( +-  2.46% )  (49.84%)
           319,993      LLC-load-misses           #   11.02% of all LL-cache hits     ( +-  3.77% )  (49.86%)

       2.075448587 seconds time elapsed                                          ( +-  0.79% )

After:
       1128.347419      task-clock (msec)         #    0.565 CPUs utilized            ( +-  0.66% )
               405      context-switches          #    0.359 K/sec                    ( +-  1.19% )
                 5      cpu-migrations            #    0.004 K/sec                    ( +- 11.08% )
             7,116      page-faults               #    0.006 M/sec                    ( +-  0.62% )
     4,516,708,282      cycles                    #    4.003 GHz                      ( +-  0.07% )  (49.82%)
    12,779,468,420      instructions              #    2.83  insn per cycle           ( +-  0.08% )  (62.39%)
     2,220,762,614      branches                  # 1968.155 M/sec                    ( +-  0.24% )  (62.69%)
         7,051,750      branch-misses             #    0.32% of all branches          ( +-  1.40% )  (62.84%)
     3,240,486,900      L1-dcache-loads           # 2871.888 M/sec                    ( +-  0.16% )  (62.66%)
        27,834,234      L1-dcache-load-misses     #    0.86% of all L1-dcache hits    ( +-  0.88% )  (62.61%)
         2,564,188      LLC-loads                 #    2.273 M/sec                    ( +-  2.12% )  (49.63%)
           312,293      LLC-load-misses           #   12.18% of all LL-cache hits     ( +-  2.11% )  (49.75%)

       1.995748820 seconds time elapsed                                          ( +-  0.16% )

arichardson committed:
[CHERI-MIPS] Check register branch (jr/jalr) targets without a helper call

This appears to be marginally slower than using the helper, but I will
commit a follow-up change that should improve this:

With helper:
       1124.972818      task-clock (msec)         #    0.563 CPUs utilized            ( +-  0.80% )
               398      context-switches          #    0.354 K/sec                    ( +-  1.17% )
                 3      cpu-migrations            #    0.003 K/sec                    ( +- 38.01% )
             7,221      page-faults               #    0.006 M/sec                    ( +-  1.18% )
     4,503,492,469      cycles                    #    4.003 GHz                      ( +-  0.34% )  (49.82%)
    12,746,394,936      instructions              #    2.83  insn per cycle           ( +-  0.32% )  (62.34%)
     2,220,269,959      branches                  # 1973.621 M/sec                    ( +-  0.40% )  (62.52%)
         7,030,906      branch-misses             #    0.32% of all branches          ( +-  0.78% )  (62.72%)
     3,234,490,689      L1-dcache-loads           # 2875.172 M/sec                    ( +-  0.21% )  (62.77%)
        27,786,841      L1-dcache-load-misses     #    0.86% of all L1-dcache hits    ( +-  0.36% )  (62.61%)
         2,521,631      LLC-loads                 #    2.242 M/sec                    ( +-  1.86% )  (49.84%)
           314,998      LLC-load-misses           #   12.49% of all LL-cache hits     ( +-  2.36% )  (49.72%)

       1.998681079 seconds time elapsed                                          ( +-  0.23% )

Without:
       1134.518390      task-clock (msec)         #    0.563 CPUs utilized            ( +-  0.72% )
               403      context-switches          #    0.355 K/sec                    ( +-  0.82% )
                 5      cpu-migrations            #    0.005 K/sec                    ( +- 23.08% )
             7,334      page-faults               #    0.006 M/sec                    ( +-  0.38% )
     4,528,989,926      cycles                    #    3.992 GHz                      ( +-  0.23% )  (49.88%)
    12,678,197,730      instructions              #    2.80  insn per cycle           ( +-  0.25% )  (62.60%)
     2,213,677,055      branches                  # 1951.204 M/sec                    ( +-  0.37% )  (62.74%)
         6,897,188      branch-misses             #    0.31% of all branches          ( +-  0.88% )  (62.84%)
     3,210,027,968      L1-dcache-loads           # 2829.419 M/sec                    ( +-  0.32% )  (62.71%)
        28,743,136      L1-dcache-load-misses     #    0.90% of all L1-dcache hits    ( +-  1.06% )  (62.51%)
         2,668,453      LLC-loads                 #    2.352 M/sec                    ( +-  1.97% )  (49.64%)
           295,481      LLC-load-misses           #   11.07% of all LL-cache hits     ( +-  5.83% )  (49.68%)

       2.014577389 seconds time elapsed                                          ( +-  0.83% )

arichardson committed:
Avoid JALR/JR dynamic checks if PCC spans the full address space

This is now a bit faster: 4,479,150,899 instead of 4,511,912,777 cycles.


       1124.005361      task-clock (msec)         #    0.554 CPUs utilized            ( +-  0.56% )
               412      context-switches          #    0.367 K/sec                    ( +-  0.99% )
                 5      cpu-migrations            #    0.004 K/sec                    ( +- 23.75% )
             7,220      page-faults               #    0.006 M/sec                    ( +-  0.61% )
     4,479,150,899      cycles                    #    3.985 GHz                      ( +-  0.17% )  (49.70%)
    12,678,267,275      instructions              #    2.83  insn per cycle           ( +-  0.27% )  (62.33%)
     2,207,617,055      branches                  # 1964.063 M/sec                    ( +-  0.18% )  (62.48%)
         7,085,827      branch-misses             #    0.32% of all branches          ( +-  1.05% )  (62.59%)
     3,211,821,304      L1-dcache-loads           # 2857.479 M/sec                    ( +-  0.18% )  (62.83%)
        27,558,280      L1-dcache-load-misses     #    0.86% of all L1-dcache hits    ( +-  0.85% )  (62.73%)
         2,602,280      LLC-loads                 #    2.315 M/sec                    ( +-  2.30% )  (49.91%)
           334,191      LLC-load-misses           #   12.84% of all LL-cache hits     ( +-  2.60% )  (49.76%)

       2.029553619 seconds time elapsed                                          ( +-  0.45% )

Before:
       1122.970888      task-clock (msec)         #    0.560 CPUs utilized            ( +-  0.52% )
               414      context-switches          #    0.368 K/sec                    ( +-  1.04% )
                 4      cpu-migrations            #    0.003 K/sec                    ( +- 27.84% )
             7,199      page-faults               #    0.006 M/sec                    ( +-  0.69% )
     4,511,912,777      cycles                    #    4.018 GHz                      ( +-  0.06% )  (49.63%)
    12,808,413,883      instructions              #    2.84  insn per cycle           ( +-  0.26% )  (62.21%)
     2,220,249,275      branches                  # 1977.121 M/sec                    ( +-  0.19% )  (62.37%)
         7,158,369      branch-misses             #    0.32% of all branches          ( +-  0.59% )  (62.70%)
     3,224,884,525      L1-dcache-loads           # 2871.744 M/sec                    ( +-  0.17% )  (62.88%)
        27,945,347      L1-dcache-load-misses     #    0.87% of all L1-dcache hits    ( +-  1.11% )  (62.80%)
         2,614,200      LLC-loads                 #    2.328 M/sec                    ( +-  2.65% )  (49.96%)
           319,453      LLC-load-misses           #   12.22% of all LL-cache hits     ( +-  2.06% )  (49.65%)

       2.003604088 seconds time elapsed                                          ( +-  0.36% )

nwf reviewed (approved):
LGTM

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -85,16 +86,93 @@ static inline void generate_get_ddc_checked_gpr_plus_offset(
 {
 #ifdef TARGET_CHERI
+    // Note: PC can only be incremented since a branch exits the TB, so checking
+    // for pc_next < pcc.base should not be needed. Add a debug assertion in
```
Assuming we're not wrapping around to 0?
EDIT: Ah but that's still caught by checking if we exceed top, so long as you're careful.

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -85,16 +86,93 @@ static inline void generate_get_ddc_checked_gpr_plus_offset(
+    // case this assumption no longer holds in the future.
+    tcg_debug_assert(ctx->base.pc_next >= ctx->base.pc_first);
     // XXX: __builtin_add_overflow() for end of address space?
     if (unlikely(ctx->base.pc_next + num_bytes > ctx->base.pcc_top)) {
```
Presumably pc_next is an address, so yes, this needs to use 128-bit arithmetic or check for overflow.

arichardson committed:
Fix setting DISAS_NORETURN for PCC violations

We can only do that if the violation is unconditional (i.e. we know whether
the branch is taken or not/the target is always OOB). Add a helper
function that saves PC and sets the flag and use it for the cases where
it's possible.

jrtc27 commented on target/cheri-common/cheri-translate-utils.h:
```diff
@@ -95,7 +95,15 @@ static inline void gen_check_pcc_bounds_next_inst(DisasContext *ctx,
+        warn_report("Found ifetch access that wraps around: " TARGET_FMT_lx
+                    " + %d.", ctx->base.pc_next, num_bytes);
+        cheri_tcg_prepare_for_unconditional_exception(&ctx->base);
+        gen_raise_pcc_violation(&ctx->base, ctx->base.pc_next, num_bytes);
```
This isn't a trap. For MIPS, alignment ensures no instruction can straddle the boundary, so when wrapping each one is in bounds if base is zero. For RISC-V without C, the same applies, and with C, instructions are fetched as 16b sized and aligned parcels, so none of them can straddle, and our specification is that each is separately checked against PCC. Though, arguably, if you get into unaligned accesses, so long as every byte is in bounds it should be fine, regardless of whether you wrap.

arichardson commented:
I'll merge this now and fix the wraparound checks in a follow-up change. I have a fix for it but it depends on a follow-up PR so adding it to this one will be awkward.
